doc/developer/design/20231216_pg_source_definite_error_status.md

```diff
@@ -0,0 +1,79 @@
+# PG sources never receive definite errors
+
+## The Problem
+
+https://github.com/MaterializeInc/materialize/issues/23951
+
+PG sources never receive definite errors from the perspective of source
+statuses, e.g. if a publication gets dropped. This makes monitoring source
+health very odd––all of the subsources report the definite error in their
+statuses, but the primary source does not.
+
+This is a problem as of
+[#23746](https://github.com/MaterializeInc/materialize/pull/23746), which
+previously papered over this issue by duplicating definite errors in subsources
+to primary sources––however, this document tackles the underlying issue.
+
+This issue will also be more apparent if we ever decide to implement a version
+of [sql: change SUBSOURCES to be a flavor of tables
+#20208](https://github.com/MaterializeInc/materialize/issues/20208) because the
+definite errors could then totally disappear from the source status table.
+
+## Success Criteria
+
+Definite errors get propagated to the source status table for the primary PG
+source.
+
+## Out of Scope
+
+This issue is narrowly scoped enough that there is nothing out of scope.
+
+## Solution Proposal
+
+The fundamental issue causing this behavior is a confluence of:
+
+-   We never want to produce data for the PG source's primary collection (as we do
+    for subsources), as it is "useless" and we eventually want to remove it
+    entirely.
+-   We currently and reasonably rely on the dataflows' health output to derive
+    statuses.
+-   `build_fallible`, which runs our fallible async operators, only produces
+    health messages when the dataflow errors––however, when the dataflow errors,
+    we place it in a "suspended" state, awaiting to be restarted. This suspended
+    state is very careful in its handling of the dataflow's output frontiers,
+    ensuring they do no emit empty frontiers when dropped.
+
+    This means that we do not have a "native" mechanism to derive statuses when
+    dataflows exit with an `Ok` result--i.e. we want to exit the dataflow and
+    close all of the input capabilities.
```

Comment:
I'm not sure I followed this reasoning. The `SourceRender` trait currently requires that implementations return a collection that contains data tagged for each individual output, including the main output. This means that all of the following options are available that don't require modifying `build_fallible`:
1. Leave it up to the source implementation to emit status updates for the main source. This could be done manually or automatically by post-processing the produced stream and duplicating errors where appropriate
2. Enforce it for all source types by automatically post-processing the resulting collection and lifting some (or all) errors from the subsources to the main source.

The issue described in https://github.com/MaterializeInc/materialize/pull/23746 was that if only one of the subsources had an issue it appeared as if the whole source had an issue. However there are multiple ways of handling this too:

1. If some subsources have errored we could automatically say something like "degraded: subsource X has such and such error"
2. Only propagate an error from subsources to the main source if all subsources have errored
3. Only propagate an error from subsources to the main source if all subsources have errored AND have advanced their frontier to the empty antichain.

So to me it sounds like #23746 was overaggressive and now we have to do a lot of work to bring back some of its benefits when a smarter promotion logic would have sufficed. What do you think?

Reply:
Opened #24253 with another approach to handle this.

I don't believe we can reasonably infer the status of the primary source from its subsources. For instance, how do you differentiate between a subsource-local definite error (e.g. a schema change) which should never apply to the primary source, and others which should? It seems fragile and annoying.

I contend that if we want the primary source to display an error, we should just propagate the error we know that we want it to display to the status manager.